### ISSUE TITLE
Hash index selectivity always 0%

### ISSUE THREAD

**Issue by ile** [arangodb/arangodb#10650]
## My Environment

* __ArangoDB Version__:        3.4.8 & 3.5.3
* __Storage Engine__:             RocksDB
* __Deployment Mode__:       Single Server
* __Deployment Strategy__:   Manual Start
* __Infrastructure__:               own 
* __Operating System__:        Ubuntu 18.04
* __Disks in use__:        SSD
* __Used Package__:             Debian or Ubuntu .deb

## Component, Query & Data

__Dataset__:

I can send a smaller version of this if needed (the problem happens also if I import just a portion of this data).

__Size of your Dataset on disk__:

```
-rw------- 1 ile ile 2,1G joulu   6 13:33 usage_9366282e11c151558bdfaab4a264aa1b.data.json
-rw------- 1 ile ile  910 joulu   6 12:47 usage_9366282e11c151558bdfaab4a264aa1b.structure.json

```

## Steps to reproduce

```

127.0.0.1:8529@db> db.usage.ensureIndex({ type: "hash", fields: [ "status" ] })
{ 
  "deduplicate" : true, 
  "fields" : [ 
    "status" 
  ], 
  "id" : "usage/119278132", 
  "isNewlyCreated" : true, 
  "selectivityEstimate" : 2.4120366417310317e-7, 
  "sparse" : false, 
  "type" : "hash", 
  "unique" : false, 
  "code" : 201 
}

127.0.0.1:8529@db> db.usage.ensureIndex({ type: "hash", fields: [ "dfghdfghdg" ] })
{ 
  "deduplicate" : true, 
  "fields" : [ 
    "dfghdfghdg" 
  ], 
  "id" : "usage/119278339", 
  "isNewlyCreated" : true, 
  "selectivityEstimate" : 1.2060183208655158e-7, 
  "sparse" : false, 
  "type" : "hash", 
  "unique" : false, 
  "code" : 201 
}


```

Actual distribution (grouping query):

```
[
  {
    "status": "status1",
    "count": 7748280
  },
  {
    "status": "status2",
    "count": 543468
  }
]
```

__Problem__:

Selectivity is always almost 0 %. `status` field exists and is a string, has two possible values and every document contains it (one value or the other). `dfghdfghdg` doesn't exist.

So it doesn't matter what fields I use, if I give some non-existent field names, I get the same selectivity. *Except indexes seem to work for some fields (e.g. timestamp field).*

__Expected result__:

Selectivity should be bigger.

**Comment by ile** [comment]
I created an index for the "user id" field (string, about 100 different users in 2 GB of data).

It took a long time (15 minutes) to create but the selectivity was 0.

```
db.usage.ensureIndex({ type: "hash", fields: [ "user" ] })
{ 
  "deduplicate" : true, 
  "fields" : [ 
    "user" 
  ], 
  "id" : "usage/123811545", 
  "isNewlyCreated" : true, 
  "name" : "idx_1652785157998903296", 
  "selectivityEstimate" : 0.00000993243696242902, 
  "sparse" : false, 
  "type" : "hash", 
  "unique" : false, 
  "code" : 201 
}
```

**Comment by jsteemann** [comment]
I don't know from the above output how many documents there are in the underlying collection.
But let's assume there are 1M documents and the indexed attribute just has two distinct values as for the `status` attribute example you gave above.
Then the expected selectivity is just 2 / 1000000, i.e. 0.000002 or 2e-6.
If the collection contains 10M documents, then the selectivity is already 2e-7.

When indexing an attribute that is not present in any document, this will be treated as all documents having the same index value (`null`), and the selectivity will be 1 / number of documents.

The selectivity values used in ArangoDB are not exact in the sense of that they are 100% accurate. They are estimates, and this explains why they are a bit off the actually expected values. But still they should be good enough for most cases.

**Comment by ile** [comment]
Hmm, ok.

The reason I created this was actually left out. It was because the queries seem to be a bit slow (I get warnings), even 
though there are indexes for these fields (status, user). 1M documents is about right.

The queries are like this:

`for u in usage filter u.status == "status1" and u.user == @v0 ... ` ... and this returns or updates some documents.

These take > 10 seconds.

`2019-12-18T00:06:24Z [1045] WARNING [8bcee] {queries} slow query: 'for u in usage filter u.status == "status1" and u.user == @v0 UPDATE u WITH { status: "status2" } in usage', bind vars: {"v0":"zyx"}, took: 11.759002 s`

Maybe I'm doing something wrong. I need to check...

**Comment by jsteemann** [comment]
Can you run an explain for the query `for u in usage filter u.status == "status1" and u.user == @v0 UPDATE u WITH { status: "status2" } in usage` to check if it uses and index, and if, which one?

If it uses an index on just `status` or just `user` that will not be as good here as a combined index on `status` _and_ `user`. This is at least true for the selection part of the query. 
If the query is dominated by the costs of updating document, then using a more selective index (as the combined index likely is) will likely not help it much. The update costs are likely dominated by the size of documents plus the number and type of indexes present for the collection. The less indexes available, the cheaper the updates will become.
So it would be useful to check if there are any indexes present on that collection that are actually not needed at all, or not used very often at least. Maybe some individual indexes can be fused into combined indexes? This is just guessing from my side, as I have no idea about the underlying data & indexes.

**Comment by ile** [comment]
Yes, I should have done this earlier:

```
Query String (107 chars, cacheable: false):
 for u in usage filter u.status == "unbilled" and u.user == @v0 UPDATE u WITH { status: "testing" } 
 in usage

Execution plan:
 Id   NodeType           Est.   Comment
  1   SingletonNode         1   * ROOT
  5   CalculationNode       1     - LET #5 = { "status" : "testing" }   /* json expression */   /* const assignment */
  7   IndexNode         74203     - FOR u IN usage   /* persistent index scan, projections: `_key`, `status` */
  8   CalculationNode   74203       - LET #3 = (u.`status` == "unbilled")   /* simple expression */   /* collections used: u : usage */
  4   FilterNode        74203       - FILTER #3
  6   UpdateNode            0       - UPDATE u WITH #5 IN usage 

Indexes used:
 By   Name      Type         Collection   Unique   Sparse   Selectivity   Fields       Ranges
  7   user      persistent   usage        false    true          0.00 %   [ `user` ]   (u.`user` == "xxxyyyzzz")
  6   primary   primary      usage        true     false       100.00 %   [ `_key` ]   u

Optimization rules applied:
 Id   RuleName
  1   move-calculations-up
  2   remove-data-modification-out-variables
  3   use-indexes
  4   remove-filter-covered-by-index
  5   patch-update-statements
  6   reduce-extraction-to-projection

Write query options:
 Option                   Value
 ignoreErrors             false
 waitForSync              false
 nullMeansRemove          false
 mergeObjects             true
 ignoreDocumentNotFound   false
 readCompleteInput        false
 useIsRestore             false
 consultAqlWriteFilter    false
 exclusive                false
 overwrite                false
 ignoreRevs               true

```

When executing, it took: **7.298 s**. So not quite the > 10 s, but I got these warnings today too:

`2019-12-19T22:50:10Z [1045] WARNING [8bcee] {queries} slow query: 'for u in usage filter u.status == "unbilled" and u.user == @v0 UPDATE u WITH { status: "billing" } in usage', bind vars: {"v0":"xxxyyyzzz"}, took: 12.920840 s`

**Comment by jsteemann** [comment]
Still wondering if it's the UPDATE that is taking long or the selection part.
I would guess it is the UPDATE part. Finding the documents via the index and post-filtering them should not take too long unless the documents are huge. Still, the selection part could be improved by using a combined index on `user` and `status`.
The UPDATE part performance will be determined most by document size plus the number of indexes available for the collection. The bigger documents or the more indexes, the slower the UPDATES will be. As I don't have any further info on indexes, document sizes etc. I can only guess here.

**Comment by ile** [comment]
Ok, I have some new tests, they seem interesting.

# 1.

`for u in usage filter u.user == @v0  collect with count into c return c`

Explain:

```
Query String (71 chars, cacheable: true):
 for u in usage filter u.user == @v0  collect with count into c return c

Execution plan:
 Id   NodeType          Est.   Comment
  1   SingletonNode        1   * ROOT
  7   IndexNode       100472     - FOR u IN usage   /* persistent index scan, scan only */
  5   CollectNode          1       - COLLECT  WITH COUNT INTO c   /* count */
  6   ReturnNode           1       - RETURN c

Indexes used:
 By   Type         Collection   Unique   Sparse   Selectivity   Fields       Ranges
  7   persistent   usage        false    true          0.00 %   [ `user` ]   (u.`user` == "xyz")

Optimization rules applied:
 Id   RuleName
  1   use-indexes
  2   remove-filter-covered-by-index
  3   remove-unnecessary-calculations-2

```

Executing:

Took **77.852 ms**, i.e. practically nothing.

Result:
[
  46180
]

# 2.

for u in usage filter u.user == @v0 and u.status == "unbilled" collect with count into c return c


Explain:

```
Query String (98 chars, cacheable: true):
 for u in usage filter u.user == @v0 and u.status == "unbilled" collect with count into c return c
 

Execution plan:
 Id   NodeType             Est.   Comment
  1   SingletonNode           1   * ROOT
  7   IndexNode         2160409     - FOR u IN usage   /* persistent index scan, projections: `user` */
  8   CalculationNode   2160409       - LET #2 = (u.`user` == "xyz")   /* simple expression */   /* collections used: u : usage */
  4   FilterNode        2160409       - FILTER #2
  5   CollectNode             1       - COLLECT  WITH COUNT INTO c   /* count */
  6   ReturnNode              1       - RETURN c

Indexes used:
 By   Type         Collection   Unique   Sparse   Selectivity   Fields         Ranges
  7   persistent   usage        false    false         0.00 %   [ `status` ]   (u.`status` == "unbilled")

Optimization rules applied:
 Id   RuleName
  1   use-indexes
  2   remove-filter-covered-by-index
  3   reduce-extraction-to-projection


```

Execution:

Took **8.678 s**, so, quite a lot. Another try around 4 s.

Result:
[
  6151
]


# 3.

`for u in usage filter u.status == "unbilled" collect with count into c return c`

Explain:

```
Query String (79 chars, cacheable: true):
 for u in usage filter u.status == "unbilled" collect with count into c return c

Execution plan:
 Id   NodeType           Est.   Comment
  1   SingletonNode         1   * ROOT
  7   IndexNode       2160409     - FOR u IN usage   /* persistent index scan, scan only */
  5   CollectNode           1       - COLLECT  WITH COUNT INTO c   /* count */
  6   ReturnNode            1       - RETURN c

Indexes used:
 By   Type         Collection   Unique   Sparse   Selectivity   Fields         Ranges
  7   persistent   usage        false    false         0.00 %   [ `status` ]   (u.`status` == "unbilled")

Optimization rules applied:
 Id   RuleName
  1   use-indexes
  2   remove-filter-covered-by-index
  3   remove-unnecessary-calculations-2
```

Execution:

Took **737.121 ms**, reasonable.

Result:
[
  503118
]


---

So, it seems that I would be much better to use just "status" in the query and not the "user" at all.

Documents in usage: 8640640

**Comment by mpoeter** [comment]
The reason why 1. and 3. perform so much better is because both queries only have to read the index data, but not the actual collection (see the explain output `persistent index scan, scan only`). The second query uses the index on `status`, but it also has to read the actual document from the collection in order to filter by `user` (see the explain output `persistent index scan, projections: user`). The second run for that query performs better, because you probably have a lot of cache hits in the block cache. You should try with a combined index for `[user, status]`. And with regards to the original query it might be interesting to see the output of a profiling run.

**Comment by ile** [comment]
Yes, true. So in effect, this was "a user error". Or, I assumed a different result (which probably wasn't all that unreasonable). 

Anyway, I created two field indexes and now the query used that and the result was very fast (undex 10 ms). So, that's good. But it wasn't easy to get here. 

Maybe it would be good to stress this fact that it's important to create an index for all the fields you plan to use and not just separate indexes. Or something. Maybe the optimizer could have optimized the index usage somehow differently? 

This can be closed, thanks for the help.

**Comment by jsteemann** [comment]
@ile : I just had a look at the latest query results you posted, and my analysis is as follows:
- there have been individual (single-field indexes) on "user" and "status". The index on "user" is sparse.
- both have a very low selectivity (i.e. lots of duplicate index values in both indexes), but the index on "user" is a bit more selective.
- in 3.4, both indexes are considered "equal" by the optimizer, because the formula for the index usage costs produces the same result for both indexes. Which one is picked is more or less random then.
- in 3.5, the index on "user" is preferred over the index on "status", because the formula for index usage costs in the optimizer has been adjusted to take post-filtering into account. So in 3.5 the query should always use the index on "user". As the above results show something contrary, I guess the results are from 3.4. Can you confirm?
- in any case, a combined index on "status" and "user" (or the other way round - for this query the order of attributes does not matter) will be more selective and will be preferred by the optimizer. Both in 3.4 and 3.5.

In summary, I guess the behavior is alright for 3.5.
For 3.4 the optimizer may pick the wrong index (the one on "status" instead the one on "user"). This is not right, but the fix for it may be quite substantial. I have to check.
A combined index will be better for both versions, but it will lead to a decrease in write performance for that collection, because an additional index has to be maintained for each write operation. Anyway, for select-mostly collections adding that combined index will be beneficial.

**Comment by jsteemann** [comment]
@ile: I just verified that a code change we did for 3.5.0 fixes the problem. So 3.5.x should favor the more selective index here. A backport of the change from 3.5 to 3.4 is not easily possible, because the change builds upon other changes that have been done in the 3.5.x+ codebase but not in 3.4.x. Plus this is a very sensitive code location that affects the index selection for all queries.
I would thus prefer to recommend using 3.5.x (or higher, 3.6.0 is already around the corner) or, simpler, creating the combined index on "user" and "status".

**Comment by maxkernbach** [comment]
Hi @ile,

Did you have a chance to read jsteemann's previous comments?
Can you confirm that your shared results are from 3.4? Is the query working as expected in 3.5?

**Comment by OmarAyo** [comment]
Hi @ile 

Just to be sure, could you kindly confirm if the output is from 3.4 or 3.5 ?

Best

**Comment by Simran-B** [comment]
I'm closing this ticket as the problem was fixed in v3.5.0, a workaround for older ArangoDB versions was suggested by jsteemann and he agreed on closing this issue.
